doc/monads/readers.lean

```diff
@@ -0,0 +1,199 @@
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `bind` function under the covers,
+so that you can access the unwrapped value `x` needed for the `toString x` conversion.
+
+The important difference here to the earlier code is that `readerFunc3` and `readerFunc2` no longer
+have an **explicit** Environment input parameter that needs to be passed along all the way to
+`readerFunc1`.  Instead, the `ReaderM` monad is taking care of that for you, which gives you the
+illusion of something like global context where the context is now available to all functions that use
+the `ReaderM` monad.
+
+The above code also introduces an important idea. Whenever you learn about a monad "X", there's
+often (but not always) a `run` function to execute that monad, and sometimes some additional
+functions like `read` that interact with the monad context.
+
+You might be wondering, how does the context actually move through the `ReaderM` monad? How can you
+add an input argument to a function by modifying it's return type?  There is a special command in
+the Lean interpreter that will show you the reduced Types:
+-/
+#reduce ReaderM Environment String   -- Environment → String
+/-!
+And you can see here that this type is actually a function!  It's a function that takes an
+`Environment` as input and returns a `String`.
+
+Now, remember in Lean that a function that takes argument of type `Nat` and returns a `String` like
+`def f (a : Nat) : String` is the same as this function `def f : Nat → String`.  These are
+definitially equal.  Well this is being used by the `ReaderM` Monad to add an input argument to
+all the functions that use the `ReaderM` monad and this is why `main` is able to start things off by
+simply passing that new input argument in `readerFunc3 env`. So now that you know the implementation
+details of the `ReaderM` monad you can see that what it is doing looks very much like the original
+code we wrote at the beginning of this section, only it's taking a lot of the tedious work off your
+plate and it is creating a nice clean separation between what your pure functions are doing, and the
+global context idea that the `ReaderM` adds.
+
+## withReader
+
+One `ReaderM` function can call another with a modified version of the `ReaderM` context. You can
+use the `withReader` function from the `MonadWithReader` typeclass to do this:
+
+-/
+def readerFunc3WithReader : ReaderM Environment String := do
+  let f := (λ env => { env with user := "new user" })
```

Comment:
```suggestion
  let f env := { env with user := "new user" }
```
Unless you want to avoid this syntax extension

Reply:
There is definitely a very real conceptual overload that happens with Lean very quickly if we throw everything in the kitchen sink at the reader all at once.  I'm struggling with this myself.  For example, I didn't even know about this syntax extension.  So, I think it might be better to move that to a more advanced chapter on really cool syntax extensions that can make your code even more beautiful, or something.

Reply:
Let's go with Sebastian's suggestion.

Reply:
Hmm, no comment on my point about conceptual overload.  So I will go with this insead which sidesteps the whole issue:

```lean
def readerFunc3WithReader : ReaderM Environment String := do
  let x ← withReader (λ env => { env with user := "new user" }) readerFunc2
  return "Result: " ++ toString x
```